Add multiple version setup to scripts

AlexBMes opened the pull request:
The scripts now:

1. initialise a git repository if one is not present
2. checkout new (or existent) branch. e.g. `gcp-cloudrun-<revision_id>` 
3. setup terraform workspace. 
4. import infrastructure
5. Instruct how to make changes and commit

gbmeuk commented on mcp/setup.sh:
```diff
@@ -3,12 +3,12 @@ set -o pipefail
 entry(){
   echo ""
-  echo "Multi-Cloud-Platform Setup"
+  echo "MCP Setup"
```
Leave this as you had it and call the script mcpadm so that you don't have a command which is `setup setup`

gbmeuk commented on mcp/setup.sh:
```diff
@@ -235,6 +240,138 @@ runDeploy() {
   entry
 }
 
+runNewVersion(){
```
What about selection of a version which already exists?

AlexBMes replied:
Added that functionality now

gbmeuk commented on mcp/setup.sh:
```diff
@@ -235,6 +240,138 @@ runDeploy() {
+  done
+}
+
+cloudRunVersion(){
```
There should be the ability to pass arguments without interaction

AlexBMes replied:
I have update to remove a portion of interaction. So running: `mcpadm new-version cloudrun` would take you to the cloudrun function. 
Did you want it to be possible to have no interaction at all? So e.g. being able to pass in `-project-id=<project-id>` or is it fine being prompted for those values?

gbmeuk replied:
yes

gbmeuk commented:
Is the python binary supposed to be included?

gbmeuk commented on mcp/mcpadm.sh:
```diff
@@ -0,0 +1,596 @@
+}
+
+s3Backend(){
+  echo "  backend \"gcs\"{" >> main.tf
```
Should this be s3... And the same in the others?

gbmeuk commented on mcp/mcpadm.sh:
```diff
@@ -0,0 +1,596 @@
+  echo "    Version: $version "
+  echo ""
+
+  select item in "yes" "no" ; do
```
This leaves interaction?

gbmeuk commented on mcp/mcpadm.sh:
```diff
@@ -0,0 +1,803 @@
+        break
+      done
+    else
+      backend=${2,,}
```
unfortunately, these nice substitutions were introduced, I think, I Bash v4. Which even the latest Mac OS doesn't include. I suggest using `tr`

gbmeuk reviewed (changes requested):
Is there are reason for the pyc file?

AlexBMes commented:
I updates the script to change the `${2,,}` style to `tr [:upper:] [:lower:]`.

I don't know how to remove the pyc file. That was added in one commit ages ago, and seemed to update when running the unit tests occasionally. 
It is not in any of the directories in this branch through, so I don't know how to remove it. I tried adding `*.pyc` to `.gitignore` but that didnt seem to remove it.

gbmeuk commented:
make setup print what the path to the created file and have some comment about updating the values afterwards because the values of the backends are all empty but we ask if the user is happy with them

gbmeuk commented:
I would expect `./mcpadm.sh --help` to return the help menu, not the interactive one

gbmeuk commented:
`source = "github.com/mesoform/terraform-infrastructure-modules//mcp"` should be the full URL and get rid of the extra forward slash.

Also make sure the readme is clear that people shouldn't be cloning this repository but instead downloading the script file and running it in `terraform/` or `deploy/` or some other directory in their application repository

AlexBMes commented:
> `source = "github.com/mesoform/terraform-infrastructure-modules//mcp"` should be the full URL and get rid of the extra forward slash.

I think you are meant to do `source = "github.com/..."` [(docs)](https://www.terraform.io/docs/language/modules/sources.html#github)

> the values of the backends are all empty but we ask if the user is happy with them

I had made it so that if doing a tag is forgotten, the user is prompted to fill in the missing value, so it shouldn't be empty. 
e.g. running `./mcpadm.sh setup gcs -bucket=bucket-name` is forgetting `-prefix=some-prefix'. So user will have to enter prefix to complete, even if `-auto-approve` flag is set.
Would you prefer these to just be left blank if using auto-approve?

gbmeuk commented:
> I think you are meant to do `source = "github.com/..."` [(docs)](https://www.terraform.io/docs/language/modules/sources.html#github)

I see. Okay. 

> > the values of the backends are all empty but we ask if the user is happy with them
> 
>
> Would you prefer these to just be left blank if using auto-approve?
I was just referring to the interactive mode

AlexBMes commented:
* Updated the script to also go to help if `--help` is run. 
* Added setup instructions in readme
* Set provider version requirements to ranges:
  * Kubernetes: `>=2.0.0, ~>2.0.2` - will allow for versions bigger 2.0.0 and less than 2.1.0
  * Google Cloud (+beta): `>=3.55.0, <4.0.0` - can be bigger than 3.55.0 but not the next major version
